src/fsharp/ParseAndCheckInputs.fs

```diff
@@ -755,6 +756,115 @@ let GetInitialTcState(m, ccuName, tcConfig: TcConfig, tcGlobals, tcImports: TcIm
       tcsRootImpls = Zset.empty qnameOrder
       tcsCcuSig = Construct.NewEmptyModuleOrNamespaceType Namespace }
 
+let mkDummyParameterVal name attribs ty =
+    Construct.NewVal(
+        name, range0, None, ty, ValMutability.Immutable, false, None, taccessPublic,
+        ValRecursiveScopeInfo.ValNotInRecScope, None, ValBaseOrThisInfo.NormalVal, attribs, ValInline.Never,
+        XmlDoc.Empty, false, false, false, false, false, false, None, ParentNone)
+
+let rec createDummyModuleOrNamespaceExpr (g: TcGlobals) (mty: ModuleOrNamespaceType) =
+
+    let dummyValAsBinding (v: Val) =
+        let dummyExpr =
+            // It does not matter what this expression is as it will never get checked or emitted.
+            let retDummyExpr = Expr.Op(TOp.Return, [], [], range0)
+
+            if isFunTy g v.Type || isForallFunctionTy g v.Type then
+                match v.ValReprInfo with
+                | Some valReprInfo ->
+                    let memberFlags =
+                        match v.MemberInfo with
+                        | Some memberInfo -> memberInfo.MemberFlags
+                        | _ ->
+                            {
+                                SynMemberFlags.IsInstance = false
+                                IsDispatchSlot = false
+                                IsOverrideOrExplicitImpl = false
+                                IsFinal = false
+                                MemberKind = SynMemberKind.Member
+                            }
+
+                    let numEnclosingTypars = CountEnclosingTyparsOfActualParentOfVal v
+                    let typars, _, curriedArgInfos, retTy, _retInfo = GetMemberTypeInMemberForm g memberFlags valReprInfo numEnclosingTypars v.Type v.Range
+                    let retTy =
+                        retTy
+                        |> Option.defaultValue g.unit_ty
+
+                    let valParams =
+                        curriedArgInfos
+                        |> List.map (fun argInfos ->
+                            argInfos
+                            |> List.map (fun (ty, argInfo) ->
+                                let name =
+                                    argInfo.Name
+                                    |> Option.map (fun x -> x.idText)
+                                    |> Option.defaultValue ""
+                                mkDummyParameterVal name argInfo.Attribs ty
+                            )
+                        )
+                    
+                    if valParams.IsEmpty || (valParams.Length = 1 && valParams.Head.IsEmpty) then
+                        // We have to create a lambda like this as `mkMemberLambdas` will throw if it is passed
+                        // a single empty curried argument list.
+                        Expr.Lambda(newUnique(), None, None, [], retDummyExpr, range0, retTy)
```

Comment:
This could have a potential bug as it's not taking into account the `typars`. Though, I don't think there is a case where we have top-level forall functions; but just in case we should handle it.